pilot/pkg/model/push_context.go

```diff
@@ -945,14 +947,13 @@ func (ps *PushContext) initSidecarScopes(env *Environment) error {
 		}
 	}
 
-	// build sidecar scopes for other namespaces that dont have a sidecar CRD object.
+	// build sidecar scopes for namespaces that dont have a non-workloadSelector sidecar CRD object.
 	// Derive the sidecar scope from the root namespace's sidecar object if present. Else fallback
 	// to the default Istio behavior mimicked by the DefaultSidecarScopeForNamespace function.
 	for _, nsMap := range ps.ServiceByHostnameAndNamespace {
 		for ns := range nsMap {
-			if len(ps.sidecarsByNamespace[ns]) == 0 {
-				// use the contents from the root namespace or the default if there is no root namespace
-				ps.sidecarsByNamespace[ns] = []*SidecarScope{ConvertToSidecarScope(ps, rootNSConfig, ns)}
+			if _, exist := sidecarConfigNSWithoutSelector[ns]; !exist {
+				ps.sidecarsByNamespace[ns] = append(ps.sidecarsByNamespace[ns], ConvertToSidecarScope(ps, rootNSConfig, ns))
```

Comment:
Rather than this logic, append the rootNSConfig to every ps.SidecarsByNamespace[sidecarConfig.Namespace] if the last Sidecar in that map entry's array is a sidecar with workloadSelector.
Also fix https://github.com/istio/istio/blob/master/pilot/pkg/model/push_context.go#L551 such that you return the root Sidecar config instead of the default sidecar scope, if the root is present.

Reply:
@rshriram Yes, this change is just what you said(append the rootNSConfig to every ps.SidecarsByNamespace[sidecarConfig.Namespace] if the last Sidecar in that map entry's array is a sidecar with workloadSelector **or the entry is empty**). After applying this change, I think the execution path will not go to https://github.com/istio/istio/blob/master/pilot/pkg/model/push_context.go#L551 to get a default sidecar. Is that correct for you?

Reply:
you removed the if check in the previous line. why?

Reply:
because if `len(ps.sidecarsByNamespace[ns]) == 0 ` then `_, exist := sidecarConfigNSWithoutSelector[ns]; !exist` will be true and will append a root/default config, so the previous line is safe to be removed.